base/special/expm1.jl

```diff
@@ -0,0 +1,238 @@
+@inline expm1_two_pow_k(::Type{Float64}, k) = reinterpret(Float64, UInt64(0x3ff00000+(k<<20))<<32)
+
+#this is 2^-k not 1-2f0^k
+@inline expm1_one_m_2_pow_mk(::Type{Float32}, k) = reinterpret(Float32, UInt32(0x3f800000 - (0x1000000>>k)))
+# 0x00000000 after & comes from lowword(1.0) = unsafe_trunc(UInt32, reinterpret(UInt64, 1.0))
+# @inline expm1_one_m_2_pow_mk(::Type{Float64}, k) = reinterpret(Float64, (UInt64(0x3ff00000 - (0x200000>>k))<<32)|0x00000000)
+@inline expm1_one_m_2_pow_mk(::Type{Float64}, k) = reinterpret(Float64, (UInt64(0x3ff00000 - (0x200000>>k))<<32) | unsafe_trunc(UInt32, reinterpret(UInt64, 1.0)))
+
+@inline expm1_two_pow_big_km1(T::Type{Float64}) = 8.98846567431158e307 # T(2.0)^(expm1_big_k(T) - 1)
+@inline expm1_two_pow_big_km1(T::Type{Float32}) = 1.7014118f38# T(2.0)^(expm1_big_k(T) - 1)
+
+function expm1(x::T) where T<:Union{Float32, Float64}
+    xsign = signbit(x)
+    absx = abs(x)
+
+    # filter out huge and non-finite argument
+    if absx >= expm1_huge(T) # quit early for large absolute values of x
+        if isinf(absx)
+            return ifelse(xsign, -T(1.0), x)
+        end
+        if x > expm1_overflow(T) # largest value T can take before overflowing
+            return T(Inf)
+        end
+        if xsign # x < -expm1_huge(T)
+            return -T(1.0)
+        end
+    end
+
+    # argument reduction
+    if absx > T(0.5)*expm1_ln2(T)
+        if absx < T(1.5)*expm1_ln2(T)
+            if xsign
+                hi = x + expm1_ln2_hi(T)
+                lo = -expm1_ln2_lo(T)
+                k = Int32(-1)
+            else
+                hi = x - expm1_ln2_hi(T)
+                lo = expm1_ln2_lo(T)
+                k = Int32(1)
+            end
+        else
+            # k  = round(Int32, expm1_invln2(T)*x)
+            # the next line could be replaced with the above but we don't need the
+            # safety of round, so we round by adding one half with the appropriate
+            # sign and truncate to Int32
+            k = unsafe_trunc(Int32, expm1_invln2(T)*x + ifelse(xsign, -T(0.5), 0.5))
+            hi = x - k*expm1_ln2_hi(T) # t*expm1_ln2_hi is exact here
+            lo = k*expm1_ln2_lo(T)
+        end
+        x = hi - lo
+        c  = (hi - x) - lo
+    elseif absx < expm1_underflow(T)
+        return x
+    else
+        k = Int32(0)
+    end
+    # x is now in primary range
+    hfx = T(0.5)*x
+    hfxs = x*hfx
+    r1 = T(1.0) + hfxs*expm1_p(hfxs)
+    t  = T(3.0) - r1*hfx
+    e  = hfxs*((r1 - t)/(T(6.0) - x*t))
+    if k == 0
+        return x - (x*e - hfxs) # c is 0
+    else
+        e  = (x*(e - c) - c)
+        e -= hfxs
+        if k == -1
+            return T(0.5)*(x - e) - T(0.5)
+        end
+        if k ==1
+            if x < -T(0.25)
+                return -T(2.0)*(e - (x + T(0.5)))
+            else
+                return  T(1.0) + T(2.0)*(x - e)
+            end
+        end
+        two_pow_k = expm1_two_pow_k(T, k) # T(2.0)^k but 9-10 times faster
```

Comment:
Should this be merged into `exp2`?   Maybe we should have an `exp2(T, x::Integer)` method.